### PR TITLE
Create Authorize and Place API Endpoint

### PR DESCRIPTION
**Pull request by JosephLeedy**
Resolves [META-182](https://boldapps.atlassian.net/browse/META-182)

- Retrieves authorized payments from Bold Checkout API
- Converts existing Magento Quote to an Order

[META-182]: https://boldapps.atlassian.net/browse/META-182?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Model/Order/PlaceOrder.php [hunk 1]
```diff
@@ -137,11 +181,234 @@ public function place(string $shopId, OrderDataInterface $order): ResultInterfac
...
+     */
+    public function authorizeAndPlace(string $publicOrderId, string $quoteMaskId): ResultInterface
+    {
+        $websiteId = (int)$this->storeManager->getStore()->getWebsiteId();
+        $shopId = $this->config->getShopId($websiteId) ?? '';
+
+        try {
+            $quoteId = $this->maskedQuoteIdToQuoteId->execute($quoteMaskId);
+        } catch (NoSuchEntityException) {
+            return $this->getValidationErrorResponse((string)__('Invalid quote mask ID "%1"', $quoteMaskId));
```
Comment (mwraich) [inline Model/Order/PlaceOrder.php:195]: [Clarification] I'm not familiar with internationalization in magento, are there commands we need to run to confirm these strings have been added or are these all pre-existing responses?

File: Model/Order/PlaceOrder.php [hunk 2]
```diff
@@ -137,11 +181,234 @@ public function place(string $shopId, OrderDataInterface $order): ResultInterfac
...
+             *     tender_details: array{
+             *         brand: string,
+             *         last_four: string,
+             *         bin: string,
+             *         expiration: string
+             *     },
+             *     gateway_response_data: string[]
+             * } $transaction
+             *
+             * TODO: verify that the status used for filtering is correct
```
Comment (mwraich) [inline Model/Order/PlaceOrder.php:304]: [Clarification] Has this been done?

File: Model/Order/PlaceOrder.php [hunk 3]
```diff
@@ -137,11 +181,234 @@ public function place(string $shopId, OrderDataInterface $order): ResultInterfac
...
+        );
+
+        $orderPayment->setBaseAmountPaid($firstTransaction['amount'] / 100);
+        $orderPayment->setAmountPaid($firstTransaction['amount'] / 100);
+        $orderPayment->setCcLast4($firstTransaction['tender_details']['last_four']);
+        $orderPayment->setCcType($firstTransaction['tender_details']['brand']);
+        $orderPayment->setCcExpMonth($cardExpirationMonth);
+        $orderPayment->setCcExpYear($cardExpirationYear);
+
+        $transaction->setTxnId($firstTransaction['transaction_id']);
```
Comment (mwraich) [inline Model/Order/PlaceOrder.php:367]: [NIT] I think we can extract these into separate functions to handle orderPayment, and transactions.

File: Model/Order/PlaceOrder.php [hunk 4]
```diff
@@ -137,11 +181,234 @@ public function place(string $shopId, OrderDataInterface $order): ResultInterfac
...
+        /** @noinspection PhpUnhandledExceptionInspection */
+        $transaction->setAdditionalInformation('gateway', $firstTransaction['gateway']);
+        /** @noinspection PhpUnhandledExceptionInspection */
+        $transaction->setAdditionalInformation('payment_id', $firstTransaction['payment_id']);
+        $transaction->setIsClosed(1);
+
+        $orderData->setQuoteId((int)$quoteId);
+        $orderData->setPublicId($publicOrderId);
+        $orderData->setPayment($orderPayment);
+        $orderData->setTransaction($transaction);
```
Comment (mwraich) [inline Model/Order/PlaceOrder.php:378]: [Clarification] Is this where we'd also add customer data?

**Review by mwraich — APPROVED** [review]
Minor nits, but overall looks good. Nothing blocking. 
I'm also not familiar with PHP, but I've tested locally and it's working as expected on my end. Haven't been able to reproduce the 422 errors but looks like they should pass along the info as expected.

**Review by nicolenorman — APPROVED** [review]

File: Model/Order/PlaceOrder.php (hunk 1 above)
Comment (JosephLeedy) [reply Model/Order/PlaceOrder.php:195]: Technically, you do not need to do anything to translate anything in Magento, other than wrapping a string with the `__()` function. (Magento will look for a translation if one is available and if not it'll just spit out the raw string.) Best practice, however, is the generate a base translation dictionary for US English and include that with the rest of your extension code. I _did_ point out to Dave, however, that we haven't done this for any of our extensions yet.

File: Model/Order/PlaceOrder.php (hunk 2 above)
Comment (JosephLeedy) [reply Model/Order/PlaceOrder.php:304]: No, that will require testing of the error cases (use one of the credit card numbers that results in a failed payment).

File: Model/Order/PlaceOrder.php (hunk 4 above)
Comment (JosephLeedy) [reply Model/Order/PlaceOrder.php:378]: No, I'm just manually constructing the same object that we passed to the previous `place()` endpoint above.

File: Model/Order/PlaceOrder.php (hunk 2 above)
Comment (JosephLeedy) [reply Model/Order/PlaceOrder.php]: Fixed.

**Pull request merged**